pkg/controller/utils/clusterdeploymentbuilder.go

```diff
@@ -0,0 +1,111 @@
+// given a specific checksum function and an existing cluster deployment.
+func NewClusterDeploymentBuilderWithClusterDeployment(checksumFunc clusterDeploymentChecksumFunc, clusterDeployment *hivev1.ClusterDeployment) *ClusterDeploymentBuilder {
+	return &ClusterDeploymentBuilder{
+		ClusterDeployment: clusterDeployment,
+		ChecksumFunc:      checksumFunc,
+	}
+}
+
+// WithName sets the name of a cluster deployment
+func (n *ClusterDeploymentBuilder) WithName(name string) *ClusterDeploymentBuilder {
+	n.ClusterDeployment.Name = name
+	return n
+}
+
+// WithNamePostfix appends the given postfix to the name of the cluster deployment.
+func (n *ClusterDeploymentBuilder) WithNamePostfix(postfix string) *ClusterDeploymentBuilder {
+	n.ClusterDeployment.Name = n.ClusterDeployment.Name + "-" + postfix
+	return n
+}
+
+// WithNamespace sets the namespace of the cluster deployment
+func (n *ClusterDeploymentBuilder) WithNamespace(namespace string) *ClusterDeploymentBuilder {
+	n.ClusterDeployment.Namespace = namespace
+	return n
+}
+
+// WithBackupChecksum sets the backup checksum for the clusterdeployment using the preveiously set checksum function.
+func (n *ClusterDeploymentBuilder) WithBackupChecksum() *ClusterDeploymentBuilder {
+	// In order for the checksum to be correct:
+	// * these two lines MUST be next to eachother
+	// * the checksum needs to be calculated AFTER all other changes have been made
+	checksum, err := n.ChecksumFunc(n.ClusterDeployment)
+
+	if err != nil {
+		n.Errors = append(n.Errors, err)
+		return n
+	}
+
+	n.WithAnnotationsSet().ClusterDeployment.ObjectMeta.Annotations[LastBackupAnnotation] = checksum
+	return n
+}
+
+// WithAnnotationsSet sets the clusterdeployment's annotations to a valid map but ONLY if it's currently set to nil.
+func (n *ClusterDeploymentBuilder) WithAnnotationsSet() *ClusterDeploymentBuilder {
+	// Only set if Nil (don't wipe out existing)
+	if n.ClusterDeployment.Annotations == nil {
+		n.ClusterDeployment.Annotations = map[string]string{}
+	}
+	return n
+}
+
+// WithEmptyAnnotations empties out the cluster deployment's annotations. If nil, then it creates an empty annotation map.
+func (n *ClusterDeploymentBuilder) WithEmptyAnnotations() *ClusterDeploymentBuilder {
+	// Wipe out existing so that it's actually empty.
+	n.ClusterDeployment.Annotations = map[string]string{}
+	return n
+}
+
+// WithEmptyBackupChecksum ensures that the cluster deployment's backup checksum is not set.
+// It attempts to emulate kube annotation behavior.
+func (n *ClusterDeploymentBuilder) WithEmptyBackupChecksum() *ClusterDeploymentBuilder {
+	// Case where annotations don't exist, so checksum doesn't exist.
+	if n.ClusterDeployment.ObjectMeta.Annotations == nil {
+		return n
+	}
+
+	// Case where Annotations exist, but checksum annotation doesn't exist.
+	if _, ok := n.ClusterDeployment.ObjectMeta.Annotations[LastBackupAnnotation]; !ok {
```

Comment:
This check is not needed. The subsequent delete will be a no-op if the annotation is not in the map.